### DISCUSSION TITLE
Support inline mermaidjs rendering

### DISCUSSION THREAD

**Discussion by damageboy** [outline/outline#1364] [General]
mermaidjs is a very popular text to: Flow Chart, Sequence Diagram, Class Diagram, State Diagram, Gantt Chart, ER Diagram
generator, capable of rendering all of the above outof markdown friendly definitions.

For the most part, mermaid users write simple definitions like below:

```mermaidjs
sequenceDiagram
	Alice->>+John: Hello John, how are you?
	Alice->>+John: John, can you hear me?
	John-->>-Alice: Hi Alice, I can hear you!
	John-->>-Alice: I feel great!
					
```

And have mermaid render this (The image below is live rendered from this text above!)::

[![](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5cdEFsaWNlLT4-K0pvaG46IEhlbGxvIEpvaG4sIGhvdyBhcmUgeW91P1xuXHRBbGljZS0-PitKb2huOiBKb2huLCBjYW4geW91IGhlYXIgbWU_XG5cdEpvaG4tLT4-LUFsaWNlOiBIaSBBbGljZSwgSSBjYW4gaGVhciB5b3UhXG5cdEpvaG4tLT4-LUFsaWNlOiBJIGZlZWwgZ3JlYXQhXG5cdFx0XHRcdFx0IiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)](https://mermaid-js.github.io/docs/mermaid-live-editor-beta/#/edit/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5cdEFsaWNlLT4-K0pvaG46IEhlbGxvIEpvaG4sIGhvdyBhcmUgeW91P1xuXHRBbGljZS0-PitKb2huOiBKb2huLCBjYW4geW91IGhlYXIgbWU_XG5cdEpvaG4tLT4-LUFsaWNlOiBIaSBBbGljZSwgSSBjYW4gaGVhciB5b3UhXG5cdEpvaG4tLT4-LUFsaWNlOiBJIGZlZWwgZ3JlYXQhXG5cdFx0XHRcdFx0IiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)

Supporting this in outline would really help elevate the ability to document and write concise documentation as outline users.

**Comment by tommoor** [comment]
I wonder if it could be stored in Markdown as a code block with a specific "language" then… are there permissively licensed javascript renderers for mermaid?

**Comment by damageboy** [comment]
I think that's exactly how it's done in various projects:
* [Typora](https://typora.io/)
* [VSCode Mermaid Support](https://marketplace.visualstudio.com/items?itemName=bierner.markdown-mermaid)

Complimentary screenshot:

![](https://github.com/mjbvz/vscode-markdown-mermaid/raw/master/docs/example.png)

**Comment by lumivn** [comment]
great idea. we really need this fuctions !

**Comment by jensschaerer** [comment]
What about this feature in Outline? Doesn't seem to be too complicated to realize, as the logic comes from Mermaid JS and only the rendering trigger has to be defined to render a codeblock with language "mermaid" as a div with class "mermaid" and embed the javascript code globally (could be done using npm / yarn?)

**Reply by tommoor** [reply]
Well, it's not too tricky but we don't want to embed the js globally across the app for performance or rely on loading js from the internet as many folks run air-gapped, so that part would need to be bundled and loaded on demand.

**Reply by jensschaerer** [reply]
I don't think it's very performance sensitive because it is only triggered on class="mermaid". Also it runs fully locally, so I don't think it should be a problem for the folks running air-gapped. Am I overseeing something? In my easy understanding, just add mairmaid as a dependency in NPM and there you go ;) I don't think it's a matter of size if the final JS is 10kb more in size

**Reply by tommoor** [reply]
The JS needs to be loaded only when there is a mermaid diagram in the page, I'm going to be quite strict about this – other than that would love for someone to take on the integration. It can be built into the existing `CodeFence` node integration.

**Comment by ArcticXWolf** [comment]
I would love to help and take a stab at implementing this, however I have some questions on the requirements and wanted to ask for your input (as I do not know which UX you would like @tommoor ).

1. Shall this be integrated in the CodeFence Node or better a separate one? As the standard Markdown for mermaidjs is wrapped inside the normal code block, I believe it is better to integrate it as opposed to deviate from the standard syntax.
2. All dependencies loaded in the editor/nodes/CodeFence.ts would only load on pages with a Code-Block?
3. Is it okay to load mermaidjs on pages with code nodes in general or do you want it to load ONLY on pages with a code node containing mermaid code?
4. How should it be displayed? We have two things that should be visible in the editor: the source of the diagram and the diagram. Is it better to show them side by side, or only show the diagram with a edit button or else?
5. If side by side: How are live updates handled in the outline UI? So how fast should the diagram be updated to reflect the diagram source during edit?
6. Mermaidjs is licensed under MIT, so is it fine to include for outline?

**Reply by ArcticXWolf** [reply]
Just a small prototype I hacked together, seems to work :)

![image](https://user-images.githubusercontent.com/5812215/167722262-71278553-3f41-4047-87b0-3c5e17ce0e27.png)

**Reply by jensschaerer** [reply]
Nice work @ArcticXWolf! I now see a little usability "issue" because the diagram and the code is displayed all the time. would be nice to hide the code somehow to only show the diagram and show the code when needed for editing. But for first shot, I like your idea!

**Reply by ArcticXWolf** [reply]
Yeah, this was not a submission of how it will look, just a small hack to see how hard it is to even integrate mermaidjs. UX will change for sure.

On that note, can someone explain how the Node-System (I believe from Prosemirror) and React components interact? For example in the shared/editor/nodes/Image.tsx we have the toDOM() method rendering a Prosemirror-Node, but also a call to React component(props) which ALSO renders a react component. Which one actually renders the Editor-Node and where does React get the content from (I believe the 0 in the toDOM specifies a content sink? Where is this in the React component?)?

Can someone help me or hint me to some links about the combination of Prosemirror and React?

**Reply by tommoor** [reply]
> we have the toDOM() method rendering a Prosemirror-Node

toDOM renders a HTML node, in React based components this is used for copy / paste and dragging only but not display.

> I believe the 0 in the toDOM specifies a content sink?

Correct! `0` specifies where child nodes will be rendered.

**Comment by ArcticXWolf** [comment]
Hey everyone,

I've added a new PR #3523 to track the progress on this feature. There are still some things to do / review, but it is a first draft.

**Reply by Limezy** [reply]
Wow it looks great already ! Congrats

**Reply by jensschaerer** [reply]
Wow great! can't wait to see it in an upcoming Outline release! :)

**Reply by tommoor** [reply]
Excited to see, heads up I'm on holiday at the moment so reviews are slower.

**Reply by ArcticXWolf** [reply]
No problem, the PR isnt ready yet anyway, there are still some things to check :)

**Reply by ArcticXWolf** [reply]
Okay, I just fixed most of the remaining problems, only two issues (concerning mobile and a bug when the codeblock is empty) are still there. I posted some questions on the PR and if someone can help me with this, then we are ready to merge :)

**Reply by ArcticXWolf** [reply]
PR #3523 is transferred into #3679 for the final polish :)

**Reply by jensschaerer** [reply]
can't wait to see the feature in a release of Outline ❤️

**Reply by ArcticXWolf** [reply]
It is merged! Now just waiting for the next release :)

**Reply by jensschaerer** [reply]
@tommoor As it is merged now, will there be a minor release or do we have to wait for 0.65? Can't await the feature for our documentation workflow

**Comment by Bradeskojest** [comment]
Is it possible to use callback functionality of Mermaid in Outline?
